sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,684 @@
+      checkState(getConnectionConfiguration() != null,
+          "SolrIO.read() requires a connection configuration"
+              + " to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      getConnectionConfiguration().populateDisplayData(builder);
+    }
+  }
+
+  /**  A POJO describing a replica of Solr. */
+  @AutoValue
+  abstract static class ReplicaInfo implements Serializable {
+    public abstract String coreName();
+    public abstract String coreUrl();
+    public abstract String baseUrl();
+
+    static ReplicaInfo create(Replica replica) {
+      return new AutoValue_SolrIO_ReplicaInfo(replica.getStr(ZkStateReader.CORE_NAME_PROP),
+          replica.getCoreUrl(), replica.getStr(ZkStateReader.BASE_URL_PROP));
+    }
+  }
+
+  /** A {@link BoundedSource} reading from Solr. */
+  @VisibleForTesting
+  static class BoundedSolrSource extends BoundedSource<SolrDocument> {
+
+    private final SolrIO.Read spec;
+    // replica is the info of the shard where the source will read the documents
+    @Nullable private final ReplicaInfo replica;
+
+    BoundedSolrSource(Read spec, @Nullable Replica replica) {
+      this.spec = spec;
+      this.replica = replica == null ? null : ReplicaInfo.create(replica);
+    }
+
+    @Override
+    public List<? extends BoundedSource<SolrDocument>> split(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      ConnectionConfiguration connectionConfig = spec.getConnectionConfiguration();
+      List<BoundedSolrSource> sources = new ArrayList<>();
+      try (AuthorizedSolrClient<CloudSolrClient> client = connectionConfig.createClient()) {
+        String collection = connectionConfig.getCollection();
+        final ClusterState clusterState = AuthorizedSolrClient.getClusterState(client);
+        DocCollection docCollection = clusterState.getCollection(collection);
+        for (Slice slice : docCollection.getSlices()) {
+          ArrayList<Replica> replicas = new ArrayList<>(slice.getReplicas());
+          Collections.shuffle(replicas);
+          // Load balancing by randomly picking an active replica
+          Replica randomActiveReplica = null;
+          for (Replica replica : replicas) {
+            // We need to check both state of the replica and live nodes
+            // to make sure that the replica is alive
+            if (replica.getState() == Replica.State.ACTIVE
+                && clusterState.getLiveNodes().contains(replica.getNodeName())) {
+              randomActiveReplica = replica;
+              break;
+            }
+          }
+          checkState(randomActiveReplica != null,
+              "Can not found an active replica for slice %s", slice.getName());
+          sources.add(new BoundedSolrSource(spec, randomActiveReplica));
```

Comment:
Sure, it's fine to address it in a separate PR. For now, a short comment in code acknowledging the issue is sufficient.